File: poc-ddd/src/domain/entities/product.ts
import Name from "../vos/name";
import Price from "../vos/price";
import UUID from "../vos/uuid";

export default class Product {
    private _id: UUID;
    private _name: Name;
    private _price: Price;

    constructor(name: string, price: number, id?: string) {
        this._id = !id ? UUID.create() : new UUID(id);
        this._name = new Name(name);
        this._price = new Price(price);
    }

    get id() {
        return this._id.getValue();
    }

    get name() {
        return this._name.getValue();
    }

    get price() {
        return this._price.getValue();
    }

    changeName(name: string): void {
        this._name = new Name(name);
    }

    changePrice(price: number): void {
        this._price = new Price(price);
    }
}

File: poc-arq-hexagonal/src/main/java/aristides/dev/portsadapters/product/core/entity/Product.java
package aristides.dev.portsadapters.product.core.entity;

import aristides.dev.portsadapters.product.core.vo.Name;
import aristides.dev.portsadapters.product.core.vo.Price;
import aristides.dev.portsadapters.product.core.vo.UUID;

import java.math.BigDecimal;

public class Product {
    private final UUID id;
    private Name name;
    private Price price;
    private ProductStatus status;

    public Product(String name, BigDecimal price) {
        this.id = UUID.create();
        this.name = new Name(name);
        this.price = new Price(price);
        this.status = ProductStatus.DISABLED;
    }

    public Product(String id, String name, BigDecimal price) {
        this.id = new UUID(id);
        this.name = new Name(name);
        this.price = new Price(price);
        this.status = ProductStatus.DISABLED;
    }

    public String getId() {
        return id.value();
    }

    public String getName() {
        return name.value();
    }

    public BigDecimal getPrice() {
        return price.value();
    }

    public ProductStatus getStatus() {
        return status;
    }

    public void enable() {
        if (this.price.value().compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalStateException("The price should be greater than zero to enable the product");
        }
        this.status = ProductStatus.ENABLED;
    }

    public void disable() {
        if (!this.price.value().equals(BigDecimal.ZERO)) {
            throw new IllegalStateException("The price should be zero to disable the product");
        }
        this.status = ProductStatus.DISABLED;
    }

    public void changeName(String name) {
        this.name = new Name(name);
    }

    public void changePrice(BigDecimal newPrice) {
        this.price = new Price(newPrice);
    }
}
